server/src/main/java/org/elasticsearch/cluster/coordination/ClusterBootstrapService.java

```diff
@@ -57,22 +62,82 @@
     public static final Setting<List<String>> INITIAL_MASTER_NODES_SETTING =
         Setting.listSetting("cluster.initial_master_nodes", Collections.emptyList(), Function.identity(), Property.NodeScope);
 
+    public static final Setting<TimeValue> UNCONFIGURED_BOOTSTRAP_TIMEOUT_SETTING =
+        Setting.timeSetting("discovery.unconfigured_bootstrap_timeout",
+            TimeValue.timeValueSeconds(3), TimeValue.timeValueMillis(1), Property.NodeScope);
+
     private final int initialMasterNodeCount;
     private final List<String> initialMasterNodes;
+    @Nullable
+    private final TimeValue unconfiguredBootstrapTimeout;
     private final TransportService transportService;
     private volatile boolean running;
 
     public ClusterBootstrapService(Settings settings, TransportService transportService) {
         initialMasterNodeCount = INITIAL_MASTER_NODE_COUNT_SETTING.get(settings);
         initialMasterNodes = INITIAL_MASTER_NODES_SETTING.get(settings);
+        unconfiguredBootstrapTimeout = discoveryIsConfigured(settings) ? null : UNCONFIGURED_BOOTSTRAP_TIMEOUT_SETTING.get(settings);
         this.transportService = transportService;
     }
 
+    public static boolean discoveryIsConfigured(Settings settings) {
+        return Stream.of(DISCOVERY_HOSTS_PROVIDER_SETTING, DISCOVERY_ZEN_PING_UNICAST_HOSTS_SETTING,
+            INITIAL_MASTER_NODE_COUNT_SETTING, INITIAL_MASTER_NODES_SETTING).anyMatch(s -> s.exists(settings));
+    }
+
     public void start() {
         assert running == false;
         running = true;
 
-        if (initialMasterNodeCount > 0 && transportService.getLocalNode().isMasterNode()) {
+        if (transportService.getLocalNode().isMasterNode() == false) {
+            return;
+        }
+
+        if (unconfiguredBootstrapTimeout != null) {
+            logger.info("no discovery configuration found, will perform best-effort cluster bootstrapping after [{}] " +
```

Comment:
can we avoid this message if the cluster is already bootstrapped?

Reply:
We can avoid the whole cluster bootstrap service in that case. I pushed 43fdbd00f5f2e737c215c562af1ab4bd38ca1c51.